src/ultk/effcomm/information_bottleneck/modeling.py

```diff
@@ -0,0 +1,350 @@
+    """
+    pxy = joint(pU_M, pM)
+    optim = IBOptimizer(
+        pxy,
+        betas,
+        **kwargs,
+    )
+    results = optim.get_results()
+    return results
+
+
+def get_ib_naming_model(
+    pU_M: np.ndarray,
+    pM: np.ndarray = None,
+    **bound_kwargs,
+) -> IBNamingModel:
+    """
+    Constructs an IBNamingModel by constructing the IB bound for the domain distribution P(M,U).
+
+    Args:
+        pU_M (np.ndarray): Conditional distribution of observations given meanings.
+        pM (np.ndarray, optional): Prior distribution over meanings. Defaults to None.
+        gammas (np.ndarray, optional): Range of gamma values for similarity selection. Defaults to logspace(-2, 2, 1000).
+        **bound_kwargs: Additional parameters for IB bound computation. See `compute_bound` kwargs.
+
+    Returns:
+        IBNamingModel: An IBNamingModel instance configured with the computed IB bound.
+
+    """
+    results = compute_bound(pU_M, pM, **bound_kwargs)
+
+    qW_M, complexity, accuracy, beta = zip(
+        *[
+            (res.qxhat_x, res.rate, res.accuracy, res.beta)
+            for res in results
+            if res is not None
+        ]
+    )
+
+    IB_curve = (np.array(complexity), np.array(accuracy))
+
+    naming_model = IBNamingModel(
+        pM[:, None],
+        pU_M,
+        beta,
+        IB_curve,
+        qW_M,
+    )
+
+    return naming_model
+
+
+##############################################################################
+# Integration with ULTK Language
+##############################################################################
+
+
+def encoder_to_language(
+    qW_M: np.ndarray,
+    naming_model: IBNamingModel,
+    universe: Universe,
+    words: list[str] = None,
+    name: str = None,
+    natural: bool = False,
+) -> Language:
+    """Convert a stochastic encoder to a ULTK Language using an IBNamingModel bayesian decoder.
+
+    Args:
+        qW_M (np.ndarray): A stochastic matrix where rows correspond to meanings
+            and columns correspond to words, defining the encoder.
+        naming_model (IBNamingModel): An instance of the IBNamingModel used to
+            decode the encoder into a language.
+        universe (Universe): The universe containing referents and the structure
+            in which the meanings are defined.
+        words (list[str], optional): A list of word forms to use. If None, default
+            numeric indices are used. Defaults to None.
+        name (str, optional): The name of the resulting Language. Defaults to None.
+        natural (bool, optional): Whether the resulting Language is a natural
+            language. Defaults to False.
+
+    Returns:
+        Language: The constructed Language object, where each expression maps a
+        word form to its corresponding meaning.
+    """
+
+    if words is None:
+        words = range(qW_M.shape[1])
+
+    return Language(
+        expressions=tuple(
+            [
```

Comment:
Don't need the list `[ ... ]` wrapper here, e.g. can delete this line and line285